### ISSUE TITLE
The implementation of observation with fairy conditions

### ISSUE THREAD

**Issue by vlaicu** [thomas-maeder/popeye#252]
Popeye applies the orthodox observation even when there are two fairy conditions, out of which one changes the way of capturing enemy pieces.

Example:
White: Bc5 Pe3
Black: Bg5
( 2 + 1 ) =1
Madrasi MarsCirce

Popeye gives the solution 1.Be7!= This solution should work only when Madrasi condition is specified. When Madrasi is combined with MarsCirce, the wBc5 is actually observed by bBg5 (via f8) and is not able to move, so the solution should be 1.e4!= Now bBg5 is observed by wBc5 (via c1).

I think the definition of observation implemented in Popeye should be carefully reconsidered in a future release. On request I can provide other [controversial] examples. 
Thank you very much for your understanding!

**Comment by flab03** [comment]
About your specific example, Popeye 4.65 onward gives the solution that you like (1.e4=). The "future release" that you ask for already exists.

From the changelog:
> 4.65 (2014-02-25)
> - more consistent and more correct handling of "observations" (check, Madrasi, Patrol Chess etc.)

In order to prove that observation should be reconsidered, you should provide one of your other controversial examples.

Personally, I'm curious about observation of enemy pieces vs friendly pieces, and MarsCirce vs Make&Take. Popeye 4.83 seems to work like this:

- MarsCirce observation of enemy pieces : yes
- MarsCirce observation of friendly pieces : yes
- Make&Take observation of enemy pieces : no
- Make&Take observation of friendly pieces : no

Is this how things should be? If we want "yes" everywhere, then how would Make&Take observation of friendly pieces work? Do we test if we could "capture" after a step in imitation of the friendly piece?

If we settle on a mix of "yes" and "no", then how do we justify the inconsistency?

**Comment by vlaicu** [comment]
Thank you for providing this answer. The specific problem I wanted to bring to your attention is the following:

Ivo Tominic, 2nd Prize SuperProblem.ru TT 211, 2018

Pieces
White Rc7 Bh7 Pb2 Kh2 Rd1
Black Qd8 Pd7 Se7 Bc6 Rf5 Pb3 Kc2
Stipulation H#2
Condition AntiCirce Lortap

The author’s intended solutions are the following:
1.Rf1 Bd3 2.Rf8 Bf5#
1.Bh1 Rc3 2.Ba8 Rc6#

Apparently Popeye thinks 3.K:d1/K:b2(>Ke8) is not legal because bPb3 observes bKc2. But, wait: the rebirth square of bPb3 observing the black King is c7 and it is occupied, so the black King is not actually observed by a friend unit and is free to capture. Similarly, in the second solution, why is the second black move legal? Apparently wRc3 is observed by wPb2 and therefore the black Bishop is no longer pinned. But again: the rebirth square of wPb2 observing the white Rook is c2 and it is occupied by the black King, so the white Rook is not observed by a friend unit and still pins the black Bishop.

Back to your question: how do we actually define "observation"? I also discussed this topic with Stephen Emmerson, who suggested the following definition (considering also the case of friendly pieces, not only enemy pieces):

> a unit observes an enemy unit if it would be possible for the first unit to capture the second under the following assumptions:
>  - the unit has the move
>  - kings are non-royal (i.e. checks or king-captures can be ignored for the purpose of testing whether one unit observes another) 
> - there are no fairy conditions (only the movement of pieces is relevant), or, the fairy conditions which apply shall be specified 
> - a unit observes a friendly unit if replacing the friendly unit by an enemy unit would also make it possible for the first unit to capture the second under the same assumptions.
> 
> This is quite wordy and I wonder how clear it now is. A more plain English definition might be:
> 
> _a unit observes another if it guards it (if it is a friendly unit) or threatens to capture it (if it is an enemy unit). To determine if this is so, imagine that the second unit replaced by an enemy pawn, also that the first unit's side has the move, that checks to kings can be ignored, and that no fairy conditions (or only specified conditions) are in force; and see if the first unit is able to capture the second._

The whole trouble with this definition is that it also allows several interpretations of observation: plain orthodox or fairy specific. The fairy specific observation might depend on the presence of other fairy conditions. Certain combinations of fairy conditions would be illegal - Christian Poisson, for instance, claims that the combination Madrasi + AntiCirce should be ruled out, just because it is not possible to define which condition has a priority.

Quoting from Christian this time:

> _Madrasi + Anticircé n’a aucun sens (et ne peut pas être vérifié par WinChloe!).
Dans la position PBb4,PNa5b7, le coup b4-b5 est impossible si l’on prend en compte Madrasi avant Anticircé (Madrasi puis Anticircé), et possible si l’on prend en compte Anticircé avant Madrasi (Anticircé puis Madrasi)._

So, before dealing with the initial request we should probably answer first another question: do you think it is possible to apply simultaneously several fairy conditions?

**Comment by flab03** [comment]
> But, wait: the rebirth square of bPb3 observing the black King is c7 and it is occupied, so the black King is not actually observed by a friend unit and is free to capture.

In feenschach 226, "The unbearable easiness of the fairy composition – part III", Example VI (Madrasi AntiCirce) is similar to AntiCirce Lortap, with 8 composers preferring orthodox observation vs 5. All you're doing is stating the minority opinion as fact. This is not a bug report. You also already know Thomas Maeder's opinion ("A mon avis, les renaissances Anti-Circé ne devraient pas influencer les observations. Le comportement de Popeye n’est donc pas faux. Mais le comportement de WinChloe n’est pas faux non plus.")

> do you think it is possible to apply simultaneously several fairy conditions?

You mean "simultaneously" as opposed to "one after the other"? Simultaneously is not always possible because it can cause contradictions. Even a single condition can cause problems if you try to make its output consistent with its input. Take wRa1 wRb1 in Lortap. They observe each other, blocking their ability to observe. Since they no longer observe each other, they regain their ability to observe, and the cycle repeats.

So applying conditions one after the other (and only once) seems to be the only way. The controversy is in the order of application.

**Comment by vlaicu** [comment]
I agree this is not a bug report, but it is still an issue which needs to be addressed. Therefore, IMHO, the result in the majority vs minority debate from feenschach has no particular relevance for the consistency of Popeye's approach.

In your example wRa1 wRb1 in Lortap, my thought is different: as both Rooks observe each other, they can't capture an enemy piece. But that doesn't mean **they lose the capability of observing friend pieces** - otherwise Lortap's definition would contain an undecidable ambiguity. The most intuitive counter-examples is wRa1 bRb1 in Madrasi: although both Rooks are paralyzed, they don't lose their power to paralyze other Rooks (e.g. wRh1 and bRa8).

In my previous comment, I was inspired by Hans Gruber, who provided another example, in order to clarify how observation should be regarded: 

> To observe is an expression for something like a geometrical description of which fields a piece could enter if there were no restrictions concerning (self)check etc. But it includes the obedience to other fairy conditions. So, in AntiCirce, wRa3 doesn’t observe en enemy piece at square a5 if field a1 is occupied.

I think his view is consistent with _Beobachtung_ definition from Die Schwalbe lexikon.

However, there is still a problem related to the possible treatment of [fairy] observation. My personal point of view is based on the following principles:
- When dealing with fairy conditions based on observation, the program should consider how the other used fairy conditions [might] influence the capturability of the enemy piece.
- In case the capturability or non-capturability of the enemy piece can't be effectively determined, the program should warn the end-user the fairy combination can't be consistently defined (i.e. non-deterministic combination of fairy elements).

My hope is this approach makes sense, although I am not sure it can be programmed. 

The alternate option, suggested by Stephen Emmerson, is to have different types of observation - which is also worth considering. Stephen even recommends a notation: putting between round brackets the specification of observation rules. For instance: Madrasi (AntiCirce) would mean that the observation used in Madrasi definition is altered by AntiCirce rules.

I find this approach more inclusive than mine, as it will allow more combinations without restricting composers' freedom. The only possible drawback I foresee is related to the potential confusion it might produce. Let's take the example VI from the aforementioned article: do you think a novice solver understand easily what's the difference between AntiCirce + Madrasi and AntiCirce + Madrasi (AntiCirce)?!
